### PR TITLE
Add assert on hash children partition count

### PR DESCRIPTION
**Pull request by duongcongtoai**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/5738

# Rationale for this change
Adding a small validation for join executors to prevent user from providing children inputs having different partition count, causing unexpected behavior

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -377,6 +370,13 @@ impl ExecutionPlan for HashJoinExec {
     ) -> Result<SendableRecordBatchStream> {
         let on_left = self.on.iter().map(|on| on.0.clone()).collect::<Vec<_>>();
         let on_right = self.on.iter().map(|on| on.1.clone()).collect::<Vec<_>>();
+        if self.mode == PartitionMode::Partitioned
+            && self.left.output_partitioning() != self.right.output_partitioning()
+        {
+            return Err(DataFusionError::Plan(format!(
```
Comment (alamb) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:376]: This is likely caused by some sort of code error, so an internal error may make more sense

```suggestion
            return Err(DataFusionError::Internal(format!(
```

Comment (alamb) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:376]: The same comment applies to the code below as well

**Review by alamb — APPROVED** [review]
Thanks @duongcongtoai  -- this is a great idea. It looks like there is a conflict that needs to be resolved, but otherwise this looks great

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @duongcongtoai

**Comment by duongcongtoai** [comment]
Test is failing, i'll check

**Comment by alamb** [comment]
I took the liberty to push a fix for the clippy CI failure: https://github.com/apache/arrow-datafusion/actions/runs/4622652579/jobs/8194716556 to your branch in https://github.com/apache/arrow-datafusion/pull/5768/commits/a29658b1140d4917c78f4d397ac77da543deee37

I also merged up from main

**Pull request merged**